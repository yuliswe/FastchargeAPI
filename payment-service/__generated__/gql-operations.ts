export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
    ID: string;
    String: string;
    Boolean: boolean;
    Int: number;
    Float: number;
    Email: any;
    NonNegativeDecimal: any;
    Timestamp: number;
};

export type GQLAccountActivity = {
    __typename?: "AccountActivity";
    amount: Scalars["String"];
    billedApp?: Maybe<GQLApp>;
    createdAt: Scalars["Timestamp"];
    description: Scalars["String"];
    pk: Scalars["ID"];
    reason: GQLAccountActivityReason;
    settleAt: Scalars["Timestamp"];
    status?: Maybe<GQLAccountActivityStatus>;
    stripeTransfer?: Maybe<GQLStripeTransfer>;
    type: GQLAccountActivityType;
    usageSummary?: Maybe<GQLUsageSummary>;
};

export enum GQLAccountActivityIndex {
    IndexByStatusSettleAtOnlyPk = "indexByStatus_settleAt__onlyPK",
}

export enum GQLAccountActivityReason {
    ApiMinMonthlyCharge = "api_min_monthly_charge",
    ApiMinMonthlyChargeUpgrade = "api_min_monthly_charge_upgrade",
    ApiPerRequestCharge = "api_per_request_charge",
    FastchargeapiPerRequestServiceFee = "fastchargeapi_per_request_service_fee",
    Payout = "payout",
    PayoutFee = "payout_fee",
    Topup = "topup",
}

export enum GQLAccountActivityStatus {
    Pending = "pending",
    Settled = "settled",
}

export enum GQLAccountActivityType {
    Credit = "credit",
    Debit = "debit",
}

export type GQLAccountHistory = {
    __typename?: "AccountHistory";
    closingBalance: Scalars["String"];
    closingTime: Scalars["Timestamp"];
};

export type GQLApp = {
    __typename?: "App";
    deleteApp: GQLApp;
    description?: Maybe<Scalars["String"]>;
    endpoints: Array<GQLEndpoint>;
    gatewayMode: GQLGatewayMode;
    homepage?: Maybe<Scalars["String"]>;
    name: Scalars["String"];
    owner: GQLUser;
    pricingPlans: Array<GQLPricing>;
    repository?: Maybe<Scalars["String"]>;
    title?: Maybe<Scalars["String"]>;
    updateApp: GQLApp;
};

export type GQLAppUpdateAppArgs = {
    description?: InputMaybe<Scalars["String"]>;
    homepage?: InputMaybe<Scalars["String"]>;
    repository?: InputMaybe<Scalars["String"]>;
    title?: InputMaybe<Scalars["String"]>;
};

export enum GQLAppIndex {
    IndexByOwnerOnlyPk = "indexByOwner__onlyPK",
}

export type GQLDateRangeInput = {
    end?: InputMaybe<Scalars["Timestamp"]>;
    start?: InputMaybe<Scalars["Timestamp"]>;
};

export type GQLEndpoint = {
    __typename?: "Endpoint";
    createdAt: Scalars["Timestamp"];
    deleteEndpoint: GQLEndpoint;
    description?: Maybe<Scalars["String"]>;
    destination?: Maybe<Scalars["String"]>;
    method: GQLHttpMethod;
    path: Scalars["String"];
    pk: Scalars["String"];
    updateEndpoint: GQLEndpoint;
    updatedAt: Scalars["Timestamp"];
};

export type GQLEndpointUpdateEndpointArgs = {
    description?: InputMaybe<Scalars["String"]>;
    destination?: InputMaybe<Scalars["String"]>;
    method?: InputMaybe<GQLHttpMethod>;
    path?: InputMaybe<Scalars["String"]>;
};

export type GQLGatewayDecisionResponse = {
    __typename?: "GatewayDecisionResponse";
    allowed: Scalars["Boolean"];
    pricingPK?: Maybe<Scalars["String"]>;
    reason?: Maybe<GQLGatewayDecisionResponseReason>;
    userPK?: Maybe<Scalars["String"]>;
};

export enum GQLGatewayDecisionResponseReason {
    InsufficientBalance = "insufficient_balance",
    NotSubscribed = "not_subscribed",
    OwnerInsufficientBalance = "owner_insufficient_balance",
    TooManyRequests = "too_many_requests",
}

export enum GQLGatewayMode {
    Proxy = "proxy",
    Redirect = "redirect",
}

export enum GQLHttpMethod {
    Any = "ANY",
    Delete = "DELETE",
    Get = "GET",
    Head = "HEAD",
    Options = "OPTIONS",
    Patch = "PATCH",
    Post = "POST",
    Put = "PUT",
}

export type GQLMutation = {
    __typename?: "Mutation";
    createApp: GQLApp;
    createEndpoint: GQLEndpoint;
    createPricing: GQLPricing;
    createSecret: GQLSecret;
    createStripeTransfer: GQLStripeTransfer;
    createSubscription: GQLSubscribe;
    createUsageLog: GQLUsageLog;
    triggerBilling: Array<GQLUsageSummary>;
};

export type GQLMutationCreateAppArgs = {
    description?: InputMaybe<Scalars["String"]>;
    gatewayMode?: InputMaybe<GQLGatewayMode>;
    homepage?: InputMaybe<Scalars["String"]>;
    name: Scalars["String"];
    owner: Scalars["ID"];
    repository?: InputMaybe<Scalars["String"]>;
    title?: InputMaybe<Scalars["String"]>;
};

export type GQLMutationCreateEndpointArgs = {
    app: Scalars["String"];
    description?: InputMaybe<Scalars["String"]>;
    destination: Scalars["String"];
    method: GQLHttpMethod;
    path: Scalars["String"];
};

export type GQLMutationCreatePricingArgs = {
    app: Scalars["String"];
    callToAction: Scalars["String"];
    chargePerRequest: Scalars["String"];
    minMonthlyCharge: Scalars["String"];
    name: Scalars["String"];
};

export type GQLMutationCreateSecretArgs = {
    description?: InputMaybe<Scalars["String"]>;
    expireAt?: InputMaybe<Scalars["Timestamp"]>;
    key: Scalars["String"];
    value: Scalars["String"];
};

export type GQLMutationCreateStripeTransferArgs = {
    currency: Scalars["String"];
    receiveAmount: Scalars["NonNegativeDecimal"];
    receiver: Scalars["ID"];
    stripeTransferId?: InputMaybe<Scalars["String"]>;
    stripeTransferObject?: InputMaybe<Scalars["String"]>;
    withdrawAmount: Scalars["NonNegativeDecimal"];
};

export type GQLMutationCreateSubscriptionArgs = {
    app: Scalars["ID"];
    pricing: Scalars["ID"];
    subscriber: Scalars["ID"];
};

export type GQLMutationCreateUsageLogArgs = {
    app: Scalars["ID"];
    path: Scalars["String"];
    pricing: Scalars["ID"];
    subscriber: Scalars["ID"];
    volume?: Scalars["Int"];
};

export type GQLMutationTriggerBillingArgs = {
    app: Scalars["ID"];
    user: Scalars["ID"];
};

export type GQLPricing = {
    __typename?: "Pricing";
    app: GQLApp;
    callToAction: Scalars["String"];
    chargePerRequest: Scalars["String"];
    deletePricing: GQLPricing;
    freeQuota: Scalars["Int"];
    minMonthlyCharge: Scalars["String"];
    mutable: Scalars["Boolean"];
    name: Scalars["String"];
    pk: Scalars["ID"];
    updatePricing: GQLPricing;
    visible: Scalars["Boolean"];
};

export type GQLPricingUpdatePricingArgs = {
    callToAction?: InputMaybe<Scalars["String"]>;
    chargePerRequest?: InputMaybe<Scalars["String"]>;
    freeQuota?: InputMaybe<Scalars["Int"]>;
    minMonthlyCharge?: InputMaybe<Scalars["String"]>;
    name?: InputMaybe<Scalars["String"]>;
    visible?: InputMaybe<Scalars["Boolean"]>;
};

export type GQLQuery = {
    __typename?: "Query";
    app: GQLApp;
    appFullTextSearch: Array<GQLApp>;
    checkUserIsAllowedForGatewayRequest: GQLGatewayDecisionResponse;
    endpoint: GQLEndpoint;
    endpoints?: Maybe<Array<Maybe<GQLEndpoint>>>;
    secret: GQLSecret;
    subscription: GQLSubscribe;
    user: GQLUser;
};

export type GQLQueryAppArgs = {
    name?: InputMaybe<Scalars["String"]>;
};

export type GQLQueryAppFullTextSearchArgs = {
    query: Scalars["String"];
};

export type GQLQueryCheckUserIsAllowedForGatewayRequestArgs = {
    app: Scalars["ID"];
    forceAwait?: InputMaybe<Scalars["Boolean"]>;
    forceBalanceCheck?: InputMaybe<Scalars["Boolean"]>;
    userEmail: Scalars["Email"];
};

export type GQLQueryEndpointArgs = {
    app?: InputMaybe<Scalars["String"]>;
    path?: InputMaybe<Scalars["String"]>;
    pk?: InputMaybe<Scalars["ID"]>;
};

export type GQLQuerySecretArgs = {
    key: Scalars["String"];
};

export type GQLQuerySubscriptionArgs = {
    app?: InputMaybe<Scalars["String"]>;
    pk?: InputMaybe<Scalars["ID"]>;
    subscriber?: InputMaybe<Scalars["ID"]>;
};

export type GQLQueryUserArgs = {
    email?: InputMaybe<Scalars["Email"]>;
    pk?: InputMaybe<Scalars["ID"]>;
};

export type GQLSecret = {
    __typename?: "Secret";
    createdAt: Scalars["Timestamp"];
    deleteSecret?: Maybe<GQLSecret>;
    expireAt?: Maybe<Scalars["Timestamp"]>;
    key: Scalars["String"];
    value: Scalars["String"];
};

export enum GQLSortDirection {
    Ascending = "ascending",
    Descending = "descending",
}

export type GQLStripePaymentAccept = {
    __typename?: "StripePaymentAccept";
    amount: Scalars["NonNegativeDecimal"];
    createdAt: Scalars["Timestamp"];
    currency: Scalars["String"];
    settlePayment: GQLStripePaymentAccept;
    status: GQLStripePaymentAcceptStatus;
    stripePaymentIntent: Scalars["String"];
    stripePaymentStatus: Scalars["String"];
    stripeSessionId: Scalars["String"];
    stripeSessionObject: Scalars["String"];
    user: GQLUser;
};

export type GQLStripePaymentAcceptSettlePaymentArgs = {
    stripeSessionObject: Scalars["String"];
};

export enum GQLStripePaymentAcceptStatus {
    Pending = "pending",
    Settled = "settled",
}

export type GQLStripeTransfer = {
    __typename?: "StripeTransfer";
    createdAt: Scalars["Timestamp"];
    currency?: Maybe<Scalars["String"]>;
    receiveAmount: Scalars["NonNegativeDecimal"];
    receiver: GQLUser;
    settleStripeTransfer: GQLStripeTransfer;
    status?: Maybe<GQLStripeTransferStatus>;
    stripeTransferId?: Maybe<Scalars["String"]>;
    stripeTransferObject?: Maybe<Scalars["String"]>;
    transferAt: Scalars["Timestamp"];
    withdrawAmount: Scalars["NonNegativeDecimal"];
};

export enum GQLStripeTransferIndex {
    IndexByStatusTransferAtOnlyPk = "indexByStatus_transferAt__onlyPK",
}

export enum GQLStripeTransferStatus {
    Failed = "failed",
    Pending = "pending",
    Transferred = "transferred",
}

export type GQLSubscribe = {
    __typename?: "Subscribe";
    app: GQLApp;
    createdAt: Scalars["Timestamp"];
    deleteSubscription?: Maybe<GQLSubscribe>;
    pk: Scalars["String"];
    pricing: GQLPricing;
    subscriber: GQLUser;
    updateSubscription?: Maybe<GQLSubscribe>;
    updatedAt: Scalars["Timestamp"];
};

export type GQLSubscribeUpdateSubscriptionArgs = {
    pricing?: InputMaybe<Scalars["ID"]>;
};

export type GQLUsageLog = {
    __typename?: "UsageLog";
    app: GQLApp;
    collectedAt: Scalars["Timestamp"];
    createdAt: Scalars["Timestamp"];
    endpoint: GQLEndpoint;
    status: Scalars["String"];
    subscriber: GQLUser;
    volume: Scalars["Int"];
};

export type GQLUsageSummary = {
    __typename?: "UsageSummary";
    app: GQLApp;
    billed: Scalars["Boolean"];
    billedAt?: Maybe<Scalars["Timestamp"]>;
    billingAccountActivity?: Maybe<GQLAccountActivity>;
    createdAt: Scalars["Timestamp"];
    subscriber: GQLUser;
    volume: Scalars["Int"];
};

export type GQLUser = {
    __typename?: "User";
    accountActivities: Array<GQLAccountActivity>;
    accountHistories: Array<GQLAccountHistory>;
    appToken: GQLUserAppToken;
    apps: Array<GQLApp>;
    author: Scalars["String"];
    balance: Scalars["String"];
    balanceLimit: Scalars["String"];
    createAppToken: GQLUserAppToken;
    createdAt: Scalars["Timestamp"];
    email: Scalars["Email"];
    pk: Scalars["ID"];
    settleAccountActivities: Array<GQLAccountActivity>;
    stripeConnectAccountId?: Maybe<Scalars["String"]>;
    stripeCustomerId?: Maybe<Scalars["String"]>;
    stripePaymentAccept: GQLStripePaymentAccept;
    subscriptions: Array<GQLSubscribe>;
    updateUser?: Maybe<GQLUser>;
    updatedAt: Scalars["Timestamp"];
    usageLogs: Array<GQLUsageLog>;
    usageSummaries: Array<GQLUsageSummary>;
};

export type GQLUserAccountActivitiesArgs = {
    dateRange?: InputMaybe<GQLDateRangeInput>;
    limit?: InputMaybe<Scalars["Int"]>;
};

export type GQLUserAccountHistoriesArgs = {
    dateRange?: InputMaybe<GQLDateRangeInput>;
    limit?: InputMaybe<Scalars["Int"]>;
};

export type GQLUserAppTokenArgs = {
    app: Scalars["ID"];
};

export type GQLUserCreateAppTokenArgs = {
    app: Scalars["ID"];
};

export type GQLUserStripePaymentAcceptArgs = {
    stripeSessionId?: InputMaybe<Scalars["String"]>;
};

export type GQLUserUpdateUserArgs = {
    author?: InputMaybe<Scalars["String"]>;
    stripeConnectAccountId?: InputMaybe<Scalars["String"]>;
    stripeCustomerId?: InputMaybe<Scalars["String"]>;
};

export type GQLUserUsageLogsArgs = {
    app?: InputMaybe<Scalars["String"]>;
    dateRange?: InputMaybe<GQLDateRangeInput>;
    limit?: InputMaybe<Scalars["Int"]>;
    path?: InputMaybe<Scalars["String"]>;
};

export type GQLUserUsageSummariesArgs = {
    app: Scalars["ID"];
    dateRange?: InputMaybe<GQLDateRangeInput>;
    limit?: InputMaybe<Scalars["Int"]>;
};

export type GQLUserAppToken = {
    __typename?: "UserAppToken";
    app: GQLApp;
    createdAt: Scalars["Timestamp"];
    deleteUserAppToken: GQLUserAppToken;
    signature: Scalars["String"];
    subscriber: GQLUser;
    token?: Maybe<Scalars["String"]>;
    updatedAt: Scalars["Timestamp"];
};

export enum GQLUserIndex {
    IndexByEmailOnlyPk = "indexByEmail__onlyPK",
}

export type GQLFulfillUserStripePaymentAcceptQueryVariables = Exact<{
    user: Scalars["ID"];
    stripeSessionId: Scalars["String"];
    stripeSessionObject: Scalars["String"];
}>;

export type GQLFulfillUserStripePaymentAcceptQuery = {
    __typename?: "Query";
    user: {
        __typename?: "User";
        stripePaymentAccept: {
            __typename?: "StripePaymentAccept";
            settlePayment: { __typename?: "StripePaymentAccept"; stripePaymentStatus: string };
        };
    };
};

export type GQLCreateStripeTransferMutationVariables = Exact<{
    userPK: Scalars["ID"];
    withdrawAmount: Scalars["NonNegativeDecimal"];
    receiveAmount: Scalars["NonNegativeDecimal"];
}>;

export type GQLCreateStripeTransferMutation = {
    __typename?: "Mutation";
    createStripeTransfer: {
        __typename?: "StripeTransfer";
        settleStripeTransfer: { __typename?: "StripeTransfer"; createdAt: number };
    };
};
