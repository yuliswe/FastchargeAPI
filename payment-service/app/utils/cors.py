import re


def cors_headers(origin: str) -> str:
    allowed_origins = [r"^http://localhost:?\d*", r"^https://.*fastchargeapi\.com"]
    if any(re.match(al, origin) for al in allowed_origins):
        return {
            "Access-Control-Allow-Origin": origin,
            "Access-Control-Allow-Headers": "*",
            "Access-Control-Allow-Methods": "*",
        }
    else:
        return {}
