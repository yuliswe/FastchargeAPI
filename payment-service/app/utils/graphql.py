from gql import Client
from gql.transport.requests import RequestsHTTPTransport
from . import signing
import os
import boto3
from botocore.auth import SigV4Auth
from botocore.awsrequest import AWSRequest
import requests
from aws_requests_auth.aws_auth import AWSRequestsAuth

from aws_requests_auth.boto_utils import BotoAWSRequestsAuth

auth = BotoAWSRequestsAuth(
    aws_host="api.iam.graphql.fastchargeapi.com",
    aws_region="us-east-1",
    aws_service="execute-api",
)

graphql_service = "https://api.iam.graphql.fastchargeapi.com"

if os.environ.get("LOCAL_GRAPHQL") == "1":
    graphql_service = "http://host.docker.internal:4000"
    auth = None


def get_graphql_client() -> Client:
    transport = RequestsHTTPTransport(
        url=graphql_service,
        headers={
            "Authorization": signing.get_payment_service_idtoken(),
            "X-User-Email": "payment-service@fastcharge-api.com",
            "Content-Type": "application/json",
            "X-Service-Name": "payment",
        },
        auth=auth,
    )
    return Client(transport=transport, fetch_schema_from_transport=True)
