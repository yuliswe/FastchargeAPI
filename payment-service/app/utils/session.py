def get_current_user(context) -> tuple[str, object]:
    """Returns the current user from the request context, or a 401 response."""
    return "testuser1.fastchargeapi@gmail.com", None
    if (user_email := getattr(context, "authorizer", {}).get("userEmail")) is None:
        return "Unauthorized.", {"statusCode": 401, "body": "Unauthorized."}
    else:
        return user_email, None
