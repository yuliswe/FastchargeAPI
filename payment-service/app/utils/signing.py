# from datetime import datetime, timedelta
# import jwt


# def get_private_key():
#     return """-----BEGIN EC PRIVATE KEY-----
# MHcCAQEEINEyilA1d68VxuH2QmIiP3+Ye6SH1/Z3/2LQc+kVZNj1oAoGCCqGSM49
# AwEHoUQDQgAE9CR7SW0cTqQBG1vxWnkjk5dO7zfvUeueXgubjSD6i6vcmHdetZ25
# /ItESQDBmX0LL2qYaPzqTJHbWKxqL+6CtA==
# -----END EC PRIVATE KEY-----
# """


# def get_public_key():
#     return """-----BEGIN PUBLIC KEY-----
# MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE9CR7SW0cTqQBG1vxWnkjk5dO7zfv
# UeueXgubjSD6i6vcmHdetZ25/ItESQDBmX0LL2qYaPzqTJHbWKxqL+6CtA==
# -----END PUBLIC KEY-----
# """


# def get_payment_service_idtoken():
#     iat = datetime.utcnow()
#     token = jwt.encode(
#         {
#             "exp": iat - timedelta(seconds=60),
#             "iat": iat,
#         },
#         get_private_key(),
#         algorithm="ES256",
#     )
#     return token
