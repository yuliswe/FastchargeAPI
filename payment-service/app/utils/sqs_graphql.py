from enum import Enum
from gql import Client
from gql.transport.requests import RequestsHTTPTransport
import boto3
from requests import Response
from json import dumps as json_dumps


class PredefinedSQSQueue(Enum):
    billing_fifo_queue = "https://sqs.us-east-1.amazonaws.com/887279901853/graphql-service-billing-queue.fifo"
    usage_log_queue = "https://sqs.us-east-1.amazonaws.com/887279901853/graphql-service-usage-log-queue"


class SQSHttpTransport(RequestsHTTPTransport):
    def __init__(
        self,
        url: str,
        queue: PredefinedSQSQueue,
        MessageDeduplicationId: str = None,
        MessageGroupId: str = None,
        **kwargs,
    ):
        super().__init__(url=url, **kwargs)
        self.MessageDeduplicationId = MessageDeduplicationId
        self.MessageGroupId = MessageGroupId
        self.queue = queue

    def connect(self):
        self.session = SQSRequestsSession(
            MessageDeduplicationId=self.MessageDeduplicationId,
            MessageGroupId=self.MessageGroupId,
            queue=self.queue,
        )


class SQSRequestsSession:
    def __init__(
        self,
        queue: PredefinedSQSQueue,
        MessageDeduplicationId: str = None,
        MessageGroupId: str = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.MessageDeduplicationId = MessageDeduplicationId
        self.MessageGroupId = MessageGroupId
        self.queue = queue
        self.client = boto3.client("sqs")

    def mount(self):
        pass

    def close(self):
        pass

    def request(
        self,
        method: str,
        url: str,
        json: object = None,
        **kwargs,
    ) -> None:
        response = self.client.send_message(
            QueueUrl=self.queue.value,
            MessageBody=json_dumps(json),
            MessageDeduplicationId=self.MessageDeduplicationId,
            MessageGroupId=self.MessageGroupId,
        )
        resp = Response()
        resp.status_code = 200
        resp._content = b'{"data": {}}'
        return resp


def get_sqs_graphql_client(
    queue: PredefinedSQSQueue,
    dedup_id: str = None,
    group_id: str = None,
) -> Client:
    """Returns a GraphQL client that uses SQS as a transport layer.

    dedup_id: Only for fifo queues. Sets the MessageDeduplicationId for the SQS
    message.

    group_id: Only for fifo queues. Sets the MessageGroupId for the SQS message.
    If the billing is PredefinedSQSQueue.billing_fifo_queue, you need to use the
    user's email as the group_id.

    Effectively, MessageDeduplicationId deduplicates messages with the same ID.
    Messages with the same MessageGroupId are processed in FIFO order.
    """
    return Client(
        transport=SQSHttpTransport(
            url="",
            queue=queue,
            MessageDeduplicationId=dedup_id,
            MessageGroupId=group_id,
        ),
        fetch_schema_from_transport=False,
    )
