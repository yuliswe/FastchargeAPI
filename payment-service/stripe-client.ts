import Stripe from "stripe";
export function getStripeClient() {
    return new Stripe(
        "sk_test_51MSgLaB24cItJ1WSPJhxovkHWjbcIcnCVmYlgCpN16acN7HUUlaLWxxcxtJT46hrvA0QHyqwWl1jxH3JeiM7ZTSi0077ZcdUOZ",
        {
            apiVersion: "2022-11-15",
        }
    );
}
