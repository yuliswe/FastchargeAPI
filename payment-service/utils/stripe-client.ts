import { getParameterFromAWSSystemsManager } from "@/functions/aws";
import { Stripe } from "stripe";
import { LambdaEventV2 } from "./LambdaContext";

let stripeClient: Stripe | null = null;
export async function getStripeClient() {
  if (stripeClient) {
    return stripeClient;
  }
  const stripeSecretKey = await getParameterFromAWSSystemsManager("payment.stripe.secret_key");
  if (!stripeSecretKey) {
    throw new Error("Missing Stripe secret key");
  }
  stripeClient = new Stripe(stripeSecretKey, { apiVersion: "2022-11-15" });
  return stripeClient;
}

/**
 * Parses and validate the signature of a webhook event received from the
 * Stripe.
 */
export const parseStripeWebhookEvent = async (lambdaEvent: LambdaEventV2): Promise<Stripe.Event> => {
  const stripeClient = await getStripeClient();
  const signature: string | undefined =
    lambdaEvent.headers["Stripe-Signature"] || lambdaEvent.headers["stripe-signature"];
  if (!signature) {
    throw new Error("Missing Stripe-Signature header");
  }
  if (!lambdaEvent.body) {
    throw new Error("Missing body");
  }

  const stripeEndpointSecret = await getParameterFromAWSSystemsManager(
    "payment.stripe.endpoint_secret.accept-stripe-payment"
  );
  if (!stripeEndpointSecret) {
    throw new Error("Missing Stripe endpoint secret");
  }

  const stripeEvent = stripeClient.webhooks.constructEvent(lambdaEvent.body, signature, stripeEndpointSecret);
  return stripeEvent;
};
