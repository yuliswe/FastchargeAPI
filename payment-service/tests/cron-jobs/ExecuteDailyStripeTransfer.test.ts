import { StripeTransfer } from "@/database/models/StripeTransfer";
import { User } from "@/database/models/User";
import { StripeTransferPK } from "@/pks/StripeTransferPK";
import { UserPK } from "@/pks/UserPK";
import { createTestStripeTransfer } from "@/tests/test-data/StripeTransfer";
import { createTestUser } from "@/tests/test-data/User";
import { baseRequestContext as context } from "@/tests/test-utils/test-utils";
import { StripeTransferStatus } from "graphql-service-apollo/__generated__/resolvers-types";
import Stripe from "stripe";
import { getTestEventBridgeEvent } from "tests/test-data/EventBridge";
import * as uuid from "uuid";
import {
  lambdaHandler as ExecuteDailyStripeTranferCronJob,
  detailType,
} from "../../cron-jobs/ExecuteDailyStripeTransfer";
import * as stripeClientModule from "../../utils/stripe-client";

describe("ExecuteDailyStripeTransfer", () => {
  let testUser: User;
  let stripeTransfer: StripeTransfer;
  let stripeTransferCreateMock: jest.SpyInstance;
  beforeEach(async () => {
    testUser = await createTestUser(context, {
      stripeConnectAccountId: uuid.v4(),
    });
    stripeTransfer = await createTestStripeTransfer(context, {
      receiver: UserPK.stringify(testUser),
      transferAt: Date.now() - 1000 * 60 * 60 * 24 * 2,
    });
    stripeTransferCreateMock = jest.fn().mockReturnValue(Promise.resolve(undefined));
    jest.spyOn(stripeClientModule, "getStripeClient").mockImplementation(() =>
      Promise.resolve({
        transfers: {
          create: stripeTransferCreateMock,
        },
      } as unknown as Stripe)
    );
  });

  test("Dry runs should not transfer money", async () => {
    const resp = ExecuteDailyStripeTranferCronJob(
      getTestEventBridgeEvent({
        detailType,
        detail: {
          dryRun: true,
        },
      })
    );
    await expect(resp).resolves.toEqual({ body: "OK", statusCode: 200 });
    const updatedStripeTransfer = await context.batched.StripeTransfer.get(StripeTransferPK.extract(stripeTransfer));
    expect(updatedStripeTransfer).toMatchObject({
      status: StripeTransferStatus.PendingTransfer,
    });
  });

  test("Transfers money", async () => {
    const resp = ExecuteDailyStripeTranferCronJob(
      getTestEventBridgeEvent({
        detailType,
        detail: {
          dryRun: false,
        },
      })
    );
    await expect(resp).resolves.toEqual({ body: "OK", statusCode: 200 });
    const updatedStripeTransfer = await context.batched.StripeTransfer.get(StripeTransferPK.extract(stripeTransfer));
    expect(updatedStripeTransfer).toMatchSnapshotExceptForProps({
      status: StripeTransferStatus.Transferred,
      receiver: UserPK.stringify(testUser),
      createdAt: expect.any(Number),
      updatedAt: expect.any(Number),
      transferAt: expect.any(Number),
    });
    expect(
      (
        stripeTransferCreateMock.mock.calls.find(
          ([{ destination }]) => destination === testUser.stripeConnectAccountId
        ) as [Stripe.TransferCreateParams]
      )[0]
    ).toMatchSnapshotExceptForProps({
      destination: testUser.stripeConnectAccountId,
      amount: 10000,
    });
  });
});
