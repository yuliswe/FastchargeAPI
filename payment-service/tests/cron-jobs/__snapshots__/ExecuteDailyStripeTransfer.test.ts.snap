// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteDailyStripeTransfer Transfers money 1`] = `
{
  "createdAt": Any<Number>,
  "currency": "usd",
  "receiveAmount": "100",
  "receiver": Any<String>,
  "status": Any<String>,
  "transferAt": Any<Number>,
  "updatedAt": Any<Number>,
  "withdrawAmount": "100",
}
`;

exports[`ExecuteDailyStripeTransfer Transfers money 2`] = `
{
  "amount": Any<Number>,
  "currency": "usd",
  "destination": Any<String>,
}
`;
